client/jetpack-connect/plans.jsx

```diff
@@ -38,65 +37,77 @@ import {
 	canCurrentUser,
 	getJetpackConnectRedirectAfterAuth,
 	isRtl,
+	isSiteAutomatedTransfer,
 	isSiteOnPaidPlan,
 } from 'state/selectors';
 import {
 	getFlowType,
-	isRedirectingToWpAdmin,
 	getSiteSelectedPlan,
 	getGlobalSelectedPlan,
 	isCalypsoStartedConnection,
 } from 'state/jetpack-connect/selectors';
 import { mc } from 'lib/analytics';
-import { isSiteAutomatedTransfer } from 'state/selectors';
+import { isJetpackSite } from 'state/sites/selectors';
 
 const CALYPSO_REDIRECTION_PAGE = '/posts/';
 const CALYPSO_PLANS_PAGE = '/plans/my-plan/';
 const JETPACK_ADMIN_PATH = '/wp-admin/admin.php?page=jetpack';
 
 class Plans extends Component {
-	static propTypes = { showJetpackFreePlan: PropTypes.bool };
-
 	static defaultProps = { siteSlug: '*' };
 
-	redirecting = false;
-
 	componentDidMount() {
-		if ( this.props.isAutomatedTransfer && ! this.redirecting && this.props.selectedSite ) {
-			this.redirecting = true;
-			this.props.goBackToWpAdmin( this.props.selectedSite.URL + JETPACK_ADMIN_PATH );
-		} else if ( this.hasPreSelectedPlan() ) {
-			this.autoselectPlan();
-		} else {
+		if ( ! this.maybeRedirect( this.props ) ) {
 			this.props.recordTracksEvent( 'calypso_jpc_plans_view', {
 				user: this.props.userId,
 			} );
 		}
 	}
 
-	componentDidUpdate() {
-		if ( this.props.isAutomatedTransfer && ! this.redirecting && this.props.selectedSite ) {
-			this.redirecting = true;
-			this.props.goBackToWpAdmin( this.props.selectedSite.URL + JETPACK_ADMIN_PATH );
+	componentWillReceiveProps = nextProps => {
+		const propsToCompare = [
+			'canPurchasePlans',
+			'flowType',
+			'hasPlan',
+			'isAutomatedTransfer',
+			'isCalypsoStartedConnection',
+			'isRequestingPlans',
+			'notJetpack',
+			'selectedPlan',
+		];
+
+		if ( ! isEqual( pick( this.props, propsToCompare ), pick( nextProps, propsToCompare ) ) ) {
+			this.maybeRedirect( nextProps );
 		}
+	};
 
-		if ( this.props.hasPlan && ! this.redirecting ) {
-			this.redirecting = true;
+	maybeRedirect = props => {
+		if ( props.isAutomatedTransfer ) {
+			this.props.goBackToWpAdmin( props.selectedSite.URL + JETPACK_ADMIN_PATH );
+			return true;
+		}
+		if ( !! props.selectedPlan ) {
```

Comment:
Is this synonymous to `hasPreSelectedPlan()`? This and the following branch are confusing…